packages/org.openapis.v3/OpenAPI.pkl

```diff
@@ -0,0 +1,200 @@
+
+//   unescaped       = %x00-2E / %x30-7D / %x7F-10FFFF
+//      ; %x2F ('/') and %x7E ('~') are excluded from 'unescaped'
+hidden expIsUnescapedCodePoint = (it) -> (it >= 0x00 && it <= 0x2E)
+    || (it >= 0x30 && it <= 0x7D)
+    || (it >= 0x7F && it <= 0x10FFFF)
+
+//   escaped         = "~" ( "0" / "1" )
+//     ; representing '~' and '/', respectively
+
+// body-reference = "body" ["#" json-pointer ]
+hidden expBodyReferenceConstraint = (it) ->
+    if (it.startsWith("#"))
+      let (suffix = it.substring(1, it.length))
+        OpenAPI.expJSONPointerConstraint.apply(suffix)
+    else
+      false
+
+// json-pointer = *( "/" reference-token )
+hidden expJSONPointerConstraint = (it) ->
+    if (it.length == 0)
+      true
+    else if (it.startsWith("/"))
+      let (suffix = it.substring(1, it.length))
+        if (suffix.indexOfOrNull("/") != null)
+          let (pre = suffix.substring(0, suffix.indexOf("/")))
+            let (post = suffix.substring(suffix.indexOf("/"), suffix.length))
+              OpenAPI.expReferenceTokenConstraint.apply(pre)
+                && OpenAPI.expJSONPointerConstraint.apply(post)
+        else
+          OpenAPI.expReferenceTokenConstraint.apply(suffix)
+    else
+      false
+
+// reference-token = *( unescaped / escaped )
+hidden expReferenceTokenConstraint = (it) ->
+  it.codePoints.fold(Pair(true, false), (acc, c) ->
+    Pair(
+      acc.first && if (acc.second)
+          (c == 0x60 || c == 0x61) // escaped values
+        else
+          (c == 0x7E || OpenAPI.expIsUnescapedCodePoint.apply(c)), // unescaped values
+      !acc.second && c == 0x7E // only escape if we're not already escaped
+    )).first
+
+// JSON string spec: https://datatracker.ietf.org/doc/html/rfc7159#section-7
+//
+// NOTE: we don't need to match %x75 separately, because all we're doing here
+// is validating that the input is correct. Hex digits should be okay
+// regardless of whether they're escaped.
+hidden expJSONNameConstraint = (it) ->
+    it.length > 0 && it.codePoints.fold(Pair(true, false), (acc, c) ->
+        Pair(
+          acc.first &&
+            if (acc.second)
+              List(0x22, 0x5C, 0x2F, 0x62, 0x66, 0x6E, 0x72, 0x74, 0x75).contains(c) // escaped values
+            else
+              c == 0x5C || (c == 0x20 || c == 0x21 || (c >= 0x23 && c <= 0x5B) || (c >= 0x5D && c <= 0x10FFFF)), // unescaped values
+          !acc.second && c == 0x5C // only escape if we're not already escaped
+        )).first
+
+/// Component reference helper function. This reference will point to the named component in the local document's
+/// "components" fields.
+function componentRef(type: Components.ComponentType, name: String): Reference =
+  // TODO: can we throw an error if the referenced component does not exist? This would be extremely useful.
+  // We may need to check for reference consistencey in the render step.
+  new Reference {
+    `$ref` = "#/components/\(type)/\(name)"
+  }
```

Comment:
This adds an output renderer, which is nice because you can get the rendered JSON/YAML results in-language just by accessing `output.text`.

This also adds a knob for controlling the output format in case users might need to control it.

```suggestion
  }

// noinspection TypeMismatch
/// The output format to render.
///
/// OpenAPI documents can be defined in either JSON or YAML.
/// This is a Pkl-only property that is excluded from OpenAPI output.
hidden __format__: "json"|"yaml" = read?("prop:pkl.outputFormat") ?? "json"

output {
  renderer =
    if (__format__ == "json")
      new JsonRenderer {}
    else
      new YamlRenderer {}
}
```